File: src/main/java/com/adven/concordion/extensions/exam/rest/commands/RequestExecutor.java
package com.adven.concordion.extensions.exam.rest.commands;

import com.google.common.base.Splitter;
import com.jayway.restassured.response.Response;
import com.jayway.restassured.specification.RequestSpecification;
import org.concordion.api.Evaluator;

import java.util.HashMap;
import java.util.Map;

import static com.jayway.restassured.RestAssured.given;

public class RequestExecutor {
    private static final String REQUEST_EXECUTOR_VARIABLE = "#request";

    private Response response;
    private String method;
    private String url;
    private String body;
    private String type;
    private Map<String, String> headers = new HashMap();
    private String urlParams;
    private String cookies;

    private RequestExecutor() {
    }

    static RequestExecutor fromEvaluator(Evaluator evaluator) {
        return (RequestExecutor) evaluator.getVariable(REQUEST_EXECUTOR_VARIABLE);
    }

    static RequestExecutor newExecutor(Evaluator evaluator) {
        RequestExecutor variable = new RequestExecutor();
        evaluator.setVariable(REQUEST_EXECUTOR_VARIABLE, variable);
        return variable;
    }

    RequestExecutor method(String method) {
        this.method = method;
        return this;
    }

    public RequestExecutor url(String url) {
        this.url = url;
        return this;
    }

    public RequestExecutor urlParams(String params) {
        this.urlParams = params;
        return this;
    }

    public RequestExecutor type(String type) {
        this.type = type;
        return this;
    }

    Response execute() {
        RequestSpecification request = given();
        request.headers(headers);

        if (body != null) {
            request.body(body);
        }

        if (type != null) {
            request.contentType(type);
        }

        if (cookies != null) {
            cookies = cookies.trim();
            if (cookies.startsWith("{")) {
                cookies = cookies.substring(1, cookies.length() - 1);
            }
            request.cookies(Splitter.on(",").trimResults().withKeyValueSeparator("=").split(cookies));
        }

        switch (method) {
            case "GET":
                response = request.get(requestUrlWithParams());
                break;
            case "POST":
                response = request.post(url);
                break;
            case "PUT":
                response = request.put(url);
                break;
            case "DELETE":
                response = request.delete(url);
                break;
            case "PATCH":
                response = request.patch(url);
                break;
            default:
                throw new UnsupportedOperationException(method);
        }
        return response;
    }

    public RequestExecutor header(String headerName, String headerValue) {
        headers.put(headerName, headerValue);
        return this;
    }

    public RequestExecutor header(Map<String, String> headersMap) {
        headers.clear();
        headers.putAll(headersMap);
        return this;
    }

    public RequestExecutor body(String body) {
        this.body = body;
        return this;
    }

    public String responseHeader(String attributeValue) {
        return response.getHeader(attributeValue);
    }

    public String statusLine() {
        return response.statusLine();
    }

    public int statusCode() {
        return response.getStatusCode();
    }

    public String responseBody() {
        return response.body().asString();
    }

    public String requestUrlWithParams() {
        return url + (urlParams != null ? "?" + urlParams : "");
    }

    public boolean hasRequestBody() {
        return method.equals("POST") || method.equals("PUT");
    }

    public String requestMethod() {
        return method;
    }

    public boolean isGET(){
        return "GET".equals(method);
    }

    public String requestHeader(String header) {
        return headers.get(header);
    }

    public RequestExecutor cookies(String cookies) {
        this.cookies = cookies;
        return this;
    }

    public String cookies() {
        return cookies;
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/ws/RequestExecutor.kt
package com.adven.concordion.extensions.exam.ws

import io.restassured.RestAssured.given
import io.restassured.http.Method
import io.restassured.http.Method.*
import io.restassured.response.Response
import org.concordion.api.Evaluator

class RequestExecutor private constructor() {
    private lateinit var method: Method
    private lateinit var url: String
    private lateinit var response: Response
    private var multiPart: MutableList<MultiPart> = ArrayList()
    private var body: String? = null
    private var type: String? = null
    private val headers = HashMap<String, String>()
    private var urlParams: String? = null
    internal var cookies: String? = null
        private set

    internal fun method(method: Method): RequestExecutor {
        this.method = method
        return this
    }

    fun url(url: String): RequestExecutor {
        this.url = url
        return this
    }

    fun urlParams(params: String?): RequestExecutor {
        this.urlParams = params
        return this
    }

    fun type(type: String): RequestExecutor {
        this.type = type
        return this
    }

    fun header(headerName: String, headerValue: String): RequestExecutor {
        headers[headerName] = headerValue
        return this
    }

    fun header(headersMap: Map<String, String>): RequestExecutor {
        headers.clear()
        headers.putAll(headersMap)
        return this
    }

    fun body(body: String): RequestExecutor {
        this.body = body
        return this
    }

    fun multiPart(name: String, fileName: String, value: Any): RequestExecutor {
        this.multiPart.add(MultiPart(name, value, fileName))
        return this
    }

    fun cookies(cookies: String?): RequestExecutor {
        this.cookies = cookies
        return this
    }

    fun xml() = type?.contains("xml", true) ?: false
    fun xml(type: String) = type.contains("xml", true)

    fun responseHeader(attributeValue: String) = response.getHeader(attributeValue)

    fun statusLine() = response.statusLine()

    fun statusCode() = response.statusCode

    fun responseBody() = response.body().asString()

    fun requestUrlWithParams() = url + if (urlParams != null) "?$urlParams" else ""

    fun hasRequestBody() = method == POST || method == PUT

    fun requestMethod() = method.name

    fun requestHeader(header: String) = headers.get(header)

    internal fun execute(): Response {
        val request = given()
        request.headers(headers)

        body?.let { request.body(it) }
        multiPart.forEach {
            if (it.value is ByteArray) {
                request.multiPart(it.name, it.type, it.value)
            } else {
                request.multiPart(it.name, it.value, it.type)
            }
        }
        type?.let { request.contentType(it) }
        cookies?.let {
            request.cookies(
                (if (it.trim().startsWith("{")) it.substring(1, it.lastIndex) else it)
                    .split(",")
                    .map {
                        val (n, v) = it.split("=")
                        Pair(n.trim(), v.trim())
                    }.toMap()
            )
        }

        response = when (method) {
            PUT -> request.put(url)
            GET -> request.get(requestUrlWithParams())
            POST -> request.post(url)
            PATCH -> request.patch(url)
            DELETE -> request.delete(requestUrlWithParams())
            else -> throw UnsupportedOperationException(method.name)
        }
        return response
    }

    companion object {
        private const val REQUEST_EXECUTOR_VARIABLE = "#request"

        internal fun fromEvaluator(evaluator: Evaluator): RequestExecutor {
            return evaluator.getVariable(REQUEST_EXECUTOR_VARIABLE) as RequestExecutor
        }

        internal fun newExecutor(evaluator: Evaluator): RequestExecutor {
            val variable = RequestExecutor()
            evaluator.setVariable(REQUEST_EXECUTOR_VARIABLE, variable)
            return variable
        }
    }
}

class MultiPart(
        val name: String,
        val value: Any,
        val type: String? = null //for file used as name
)
